stub_uploader/metadata.py

```diff
@@ -283,23 +284,37 @@ def verify_external_req(
     upstream_distribution: Optional[str],
     _unsafe_ignore_allowlist: bool = False,  # used for tests
 ) -> None:
-    req_canonical_name = canonical_name(req.name)
+    """Verify that a non-typeshed dependency is valid.
+
+    Raise InvalidRequires if the dependency is invalid.
+    """
+
+    verify_external_req_not_in_typeshed(req)
+    verify_external_req_name(req)
+    verify_external_req_in_allowlist(req, _unsafe_ignore_allowlist)
+    verify_external_req_stubs_require_its_runtime(req, upstream_distribution)
 
+
+def verify_external_req_not_in_typeshed(req: Requirement) -> None:
+    req_canonical_name = canonical_name(req.name)
     if req_canonical_name in uploaded_packages.read():
         raise InvalidRequires(
             f"Expected dependency {req} to not be uploaded from stub_uploader"
         )
+
+
+def verify_external_req_name(req: Requirement) -> None:
     if req.name.startswith(TYPES_PREFIX):
         # technically this could be allowed, but it's very suspicious
         raise InvalidRequires(
             f"Expected dependency {req} to not start with {TYPES_PREFIX}"
         )
 
-    if upstream_distribution is None:
-        raise InvalidRequires(
-            f"There is no upstream distribution on PyPI, so cannot verify {req}"
-        )
 
+def verify_external_req_in_allowlist(
+    req: Requirement,
+    _unsafe_ignore_allowlist: bool = False,  # used for tests
```

Comment:
Let's take a page out of flake8-boolean-trap's book to help ensure this isn't used accidentally and is easily recognizable from the call site:
```suggestion
    *,
    _unsafe_ignore_allowlist: bool = False,  # used for tests
```